crates/swc_ecma_utils/src/function/function_wrapper.rs

```diff
@@ -0,0 +1,394 @@
+            left: PatOrExpr::Expr(Box::new(Expr::Ident(ref_ident.clone()))),
+            right: Box::new(self.function.take()),
+        }
+        .into_stmt();
+
+        // clone `return REF.apply(this, arguments);`
+        let return_ref_apply_stmt = function.body.as_ref().unwrap().stmts[0].clone();
+
+        let ref_fn_block_stmt = BlockStmt {
+            span: DUMMY_SP,
+            stmts: vec![assign_stmt, return_ref_apply_stmt],
+        };
+
+        let ref_decl = FnDecl {
+            ident: ref_ident,
+            declare: false,
+            function: Function {
+                span: DUMMY_SP,
+                is_async: false,
+                is_generator: false,
+                params: self.params.take(),
+                body: Some(ref_fn_block_stmt),
+                decorators: Default::default(),
+                type_params: Default::default(),
+                return_type: Default::default(),
+            },
+        };
+
+        let name_decl = FnDecl {
+            ident: name_ident,
+            function,
+            declare: false,
+        };
+
+        (name_decl, ref_decl)
+    }
+
+    ///
+    /// ```javascript
+    /// function NAME(PARAMS) {
+    ///     return REF.apply(this, arguments);
+    /// }
+    /// ```
+    fn build_function_forward(&mut self, ref_ident: Ident, name_ident: Option<Ident>) -> FnExpr {
+        let apply = Stmt::Return(ReturnStmt {
+            span: DUMMY_SP,
+            arg: Some(Box::new(ref_ident.apply(
+                DUMMY_SP,
+                Box::new(Expr::This(ThisExpr { span: DUMMY_SP })),
+                vec![quote_ident!("arguments").as_arg()],
+            ))),
+        });
+
+        FnExpr {
+            ident: name_ident,
+            function: Function {
+                span: DUMMY_SP,
+                is_async: false,
+                is_generator: false,
+                params: self.params.take(),
+                body: Some(BlockStmt {
+                    span: DUMMY_SP,
+                    stmts: vec![apply],
+                }),
+                decorators: Default::default(),
+                type_params: Default::default(),
+                return_type: Default::default(),
+            },
+        }
+    }
+}
+
+impl Into<Expr> for FunctionWrapper {
+    fn into(mut self) -> Expr {
+        // A function may be called internally recursively if it has a name
+        if let Some(name_ident) = self.function_ident.as_ref().cloned() {
+            self.build_named_expression_wrapper(name_ident)
+        }
+        // An internal recursive call has to refer to the external binding_name,
+        // It is safe to using expression wrapper.
+        else if self.params.len() > 0 || self.binding_ident.is_some() {
+            self.build_anonymous_expression_wrapper()
+        } else {
+            // Optimization: no name, no params
+            self.function
+        }
+    }
+}
+
+impl Into<(FnDecl, FnDecl)> for FunctionWrapper {
```

Comment:
This `impl` seems too opinionated compared to the public API.
What `FnDecl` it creates? And, why it returns two function declarations?
(I know because I'm looking at source code, but uses will only see `.into()`)

I think it should use a dedicated method which has a describing name.